planner/cascades/implementation_rules.go

```diff
@@ -272,3 +277,60 @@ func (r *ImplLimit) OnImplement(expr *memo.GroupExpr, reqProp *property.Physical
 	}.Init(logicalLimit.SCtx(), expr.Group.Prop.Stats, logicalLimit.SelectBlockOffset(), newProp)
 	return impl.NewLimitImpl(physicalLimit), nil
 }
+
+// ImplTopN is the implementation rule which implements LogicalTopN
+// to PhysicalTopN.
+type ImplTopN struct {
+}
+
+// Match implements ImplementationRule Match interface.
+func (r *ImplTopN) Match(expr *memo.GroupExpr, prop *property.PhysicalProperty) (matched bool) {
+	topN := expr.ExprNode.(*plannercore.LogicalTopN)
+	return plannercore.MatchItems(prop, topN.ByItems)
+}
+
+// OnImplement implements ImplementationRule OnImplement interface.
+func (r *ImplTopN) OnImplement(expr *memo.GroupExpr, reqProp *property.PhysicalProperty) (memo.Implementation, error) {
+	lt := expr.ExprNode.(*plannercore.LogicalTopN)
+	resultProp := &property.PhysicalProperty{ExpectedCnt: math.MaxFloat64}
+	topN := plannercore.PhysicalTopN{
+		ByItems: lt.ByItems,
+		Count:   lt.Count,
+		Offset:  lt.Offset,
+	}.Init(lt.SCtx(), expr.Group.Prop.Stats, lt.SelectBlockOffset(), resultProp)
+	switch expr.Group.EngineType {
+	case memo.EngineTiDB:
+		return impl.NewTiDBTopNImpl(topN), nil
+	default:
+		// TODO: return TiKVTopNImpl after we have implemented push topN down gather.
+		return nil, plannercore.ErrInternal.GenWithStack("Unsupported EngineType '%s' for TopN.", expr.Group.EngineType.String())
+	}
+}
+
+// ImplTopNAsLimit is the implementation rule which implements LogicalTopN
+// as PhysicalLimit with required order property.
+type ImplTopNAsLimit struct {
+}
+
+// Match implements ImplementationRule Match interface.
+func (r *ImplTopNAsLimit) Match(expr *memo.GroupExpr, prop *property.PhysicalProperty) (matched bool) {
+	topN := expr.ExprNode.(*plannercore.LogicalTopN)
+	_, canUseLimit := plannercore.GetPropByOrderByItems(topN.ByItems)
+	return plannercore.MatchItems(prop, topN.ByItems) && canUseLimit
```

Comment:
`canUseLimit && plannercore.MatchItems(prop, topN.ByItems)` to take the benefits of short circuit evaluation?